src/mds/MDSMap.cc

```diff
@@ -939,6 +945,7 @@ void MDSMap::decode(bufferlist::const_iterator& p)
   }
 
   if (ev >= 17) {
+    decode(max_xattr_size, p);
```

Comment:
I am concerned with this and other corresponding changes in the d/encoder where the format is updated without bumping up the version.
I understand that the logic here is to "catch up" with main where both fields are serialized under v 17, but since main isn't released yet and reef has already been released, we can't afford this.

This change should introduce version 18 with the second field, and then main should be updated to expect one field in v 17 and the second in v 18

Reply:
I'm going to take another look at the kclient and the mds changes tomorrow. Will update once done.

Reply:
OK. I did some digging around in the history where all this started. So, just quickly - reef (18.2.0) does:

```
if (version >=17) {
    decode(bal_rank_mask, p);
}
```

while main (to-be-branched-out-and-released squid-release) does

```
version = 17
encode(version, bl);
...
...
encode(max_xattr_size, bl);
encode(bal_rank_mask, bl);
```

which is the problem since 18.2.0 would decode max_xattr_size (type: uint64_t) into bal_rank_mask (type: string).

Also, the kclient does:

```
        if (mdsmap_ev >= 17) {
                /* balancer */
                ceph_decode_skip_string(p, end, bad_ext);
                /* standby_count_wanted */
                ceph_decode_skip_32(p, end, bad_ext);
                /* old_max_mds */
                ceph_decode_skip_32(p, end, bad_ext);
                /* min_compat_client */
                ceph_decode_skip_8(p, end, bad_ext);
                /* required_client_features */
                ceph_decode_skip_set(p, end, 64, bad_ext);
                ceph_decode_64_safe(p, end, m->m_max_xattr_size, bad_ext);
        } else {
                /* This forces the usage of the (sync) SETXATTR Op */
                m->m_max_xattr_size = 0;
        }
```

and the kernel version is already out in the field way before all this started (since the kclient changes were merged very early on).

So, the options we have are
- encode both fields with v17: This breaks reef releases which decodes bal_rank_mask (18.2.0 and probably the next minor release - till this fix is merged). Not an issue for kclient.
- encode max_xattr_size with v17 and bal_rank_mask with v18: This breaks 18.2.0. Not an issue for kclient.
- encode bal_rank_mask with v17 and max_xattr_size with v18: Breaks kclient.

cc @lxbsz

Reply:
cc @luis-henrix

Reply:
Yikes!  So PR #46357 wasn't merged into reef!?!? Somehow I assumed it was, although, to be honest, I didn't check -- when I saw it was merged, I assumed it was picked into reef.  That's not good.

I haven't looked again at the code that handles the `bal_rank_mask` to refresh my memory but my opinion is that, because any other option will break something else, this backport needs to fix it -- i.e. both `bal_rank_mask` and `max_xattr_size`.

In fact, if I understand it correctly, the kernel is already somewhat broken with the reef release because it will decode a value for `max_xattr_size` out of the string encoded in `bal_rank_mask`.  It's probably not a big deal, it's just as bad as not having the fix at all.  I.e. the client may try to set a huge xattr, and not get an error even if the xattr wasn't set because it was too big (see simple test-case in the origin PR).

Unfortunately, this is the fallback from taking too long merging the original PR: we agreed that the hacky solution was acceptable **if** it could be merged quickly ;-)

Reply:
OK, another thought: since the kernel client seems to be broken anyway, going with the 3rd option (`bal_rank_mask` with v17 and `max_xattr_size` with v18) may also be acceptable.  But I suspect that broken kernel clients will be around for several years, even if we provide backports for all stable kernels.

@lxbsz what do you think?

Reply:
Sorry I missed this thread.

Since we have downstream the corresponding kernel patches, then the option 2 is the best IMO.

While the option 3 will make the situation more complex.

Reply:
Fair enough. In that case, I'll send a change to the main branch for the same.

Reply:
@vshankar It's only broken in case of `reef` releases. Not the main. For the `option 3` we need to adjust kernels, including the downstream kernels of rhel 8/9, which sounds not a good idea.

Reply:
OK. In that case, let's stick to option (2).

Reply:
Oops, I feel like a being a bit late to the part but the decoding change just popped up in reef2squid upgrade testing: https://trello.com/c/NCZCeuVu/1894-wip-yuri10-testing-2023-12-05-1105-old-wip-yuri10-testing-2023-12-04-1344-old-wip-yuri10-testing-2023-11-30-0943-old-wip-yuri10#comment-657258a40cbac82ea76c6d02.

Reply:
Now that https://github.com/ceph/ceph/pull/54745 is merged in main (squid+), let's fix the decode here accordingly. @trociny 

Let's also add a note in PendingReleaseNotes for this (breakage).

Reply:
@vshankar I have cherry-picked your PR. As for PendingReleaseNote I suppose a PR should be created against the main branch first and then cherry-picked here too? Do you want me to do this?

Reply:
Thx for the update. I'll take care of PendingReleaseNote update.

Reply:
Sorry, I have to reopen this discussion. We must go with option (3). Option (2) causes upgrades from `v18.2.[01]` to fail `v18.2.2`:

`2024-02-15T01:29:59.124+0000 7f2c15711700 -1 failed to decode message of type 45 v1: End of buffer [buffer:2]`

That is with an upgraded `ceph-mgr` receiving the `FSMap` generated by a `v18.2.1` `ceph-mon`. This is expected because the monitor is sending `bal_rank_mask` for `ev >= 17` but (if 0 length) its encoded size is a 32-bit integer. The `max_xattr_size` is a `uint64_t` so attempting to decode `bal_rank_mask` as a 64-bit int will always fail `End of buffer`.

Obviously breaking the client is undesirable but upgrades **must not be broken**. The only way forward here is to change `main` to reverse the encoding according to how `v18.2.[01]` already works in the wild, fix `v18.2.2` (as this is already merged.

I will be making a notice to not merge the `quincy`/`pacific` backports.

I'll also open a ticket that we must have upgrade testing from older minor releases for the same major version (i.e. v18.2.1 -> reef HEAD).

Reply:
https://pulpito.ceph.com/pdonnell-2024-02-15_13:49:19-fs:upgrade:mds_upgrade_sequence-reef-distro-default-smithi/7561035/

Reply:
> Obviously breaking the kclient is undesirable but upgrades **must not be broken**. The only way forward here is to change `main` to reverse the encoding according to how `v18.2.[01]` already works in the wild, fix `reef` (as this PR is already merged).

So, it has really come to the point where we *cannot* continue without breaking either the upgrade or the kclient which is a pity. The kclient breakage wasn't caught for long since the encoded string was decode into 64bit integer and probably a combination of lack of test coverage. What's interesting to note is the early merge of the kclient changes esp. which bump-up the encode/decode versions and the chance of an entire different MDS change that uses the exact same encode/decode version getting merged later.

Reply:
Yeah, before when we didn't notice this situation and Jeff just applied the kclient patches before the `max_xattr_size` ceph pr, which was deferred around 1 month or longer, was ready. And just during that the `bal_rank_mask` ceph pr was merged.

As I mentioned in that PR, we will always to make sure the cephfs side PRs to get merged and then the kclient patches.